prjxray/timing.py

```diff
@@ -0,0 +1,615 @@
++--+-+           +--+-+
+   |                |
+   v                v
++--+--+          +--+--+
+|Inpin|          |Inpin|
++-----+          +-----+
+
+Note on units:
+
+The timing model operates on the following types of units:
+ - Time
+ - Resistance
+ - Capacitance
+
+For a consistent unit set, the following equation must be satisfied:
+
+1 Resistance unit * 1 Capacitance unit = 1 Time unit
+
+The SI unit set would be:
+ - Time = seconds
+ - Resistance = Ohms
+ - Capacitance = Farads
+
+However as long as the scale factors are consistent, the model will work
+with other unit combinations.  For example:
+ - Time = nanoseconds (1e-9 seconds)
+ - Resistance = milliOhms (1e-3 Ohms)
+ - Capacitance = microFarads (1e-6 Farads)
+
+(1e-3 * 1e-6) (Ohms * Farads) does equal (1e-9) seconds.
+
+"""
+import enum
+from collections import namedtuple
+
+
+class PvtCorner(enum.Enum):
+    """ Process/voltage/temperature corner definitions. """
+
+    # Corner where device operates with fastest intristic delays.
+    FAST = "FAST"
+
+    # Corner where device operates with slowest intristic delays.
+    SLOW = "SLOW"
+
+
+class IntristicDelay(namedtuple('IntristicDelay', 'min max')):
+    """ An intristic delay instance.
+
+    Represents is the intristic delay through an element (e.g. a site pin or
+    interconnect pip).
+
+    The intristic delay of an element is generally modelled at a particular
+    "corner" of a design.  The "corner" generally speaking is modelled over
+    process, voltage and temperature PVT.  The IntristicDelay object
+    reperesents the minimum or maximum delay through all instances of the
+    element at 1 corner.
+
+    Attributes
+    ----------
+
+    min : float
+        Minimum instrinsic delay (nsec)
+
+    max : float
+        Maximum instrinsic delay (nsec)
+    """
+
+
+class RcElement(namedtuple('RcElement', 'resistance capacitance')):
+    """ One part of an RcNode, embedded within an RcTree.
+
+    Attributes
+    ----------
+
+    resistance : float
+        Resistance of element
+
+    capacitance : float
+        Capacitance of element
+
+    """
+    pass
+
+
+def fast_slow_tuple_to_corners(arr):
+    """ Convert delay 4-tuple into two IntristicDelay objects.
+
+    Returns
+    -------
+
+    corners : dict of PvtCorner to IntristicDelay
+        Dictionary keys of FAST and SLOW, mapping to the instrinsic delay
+        for each corner.
+
+    """
+
+    fast_min, fast_max, slow_min, slow_max = map(float, arr)
+
+    return {
+        PvtCorner.FAST: IntristicDelay(
+            min=fast_min,
+            max=fast_max,
+        ),
+        PvtCorner.SLOW: IntristicDelay(
+            min=slow_min,
+            max=slow_max,
+        ),
+    }
+
+
+# Math models are used to represent abstract operations for the timing models.
+# This is useful for creating excel workbooks that can update dynamically, or
+# generating a model with symbolic constants to be backsolved.
+
+
+class ExcelMathModel(object):
```

Comment:
I feel like this might be better in it's own file?

Reply:
Okay